File: python/src/lessons_basics.py
"""Modules 1 to 5 - the basics.

Rule used throughout: nothing is claimed without being shown. Where an
example can run here, it runs.
"""

import ui


def lesson_01_running():
    ui.title("MODULE 1 - RUNNING PYTHON AND PRINT()")

    ui.heading("PART 1: no compile step")

    print("  A C program has to be translated into machine code before it")
    print("  can run. Python skips that: the interpreter reads your file")
    print("  and runs it directly, line by line.")
    print()
    print("    python3 program.py")
    print()
    print("  That is the whole command. No -o, no linking, no separate")
    print("  binary left behind. It also means a mistake on line 40 is not")
    print("  caught until the interpreter actually reaches line 40 -- a C")
    print("  compiler would have refused to build at all. Python trades an")
    print("  early, whole-program check for a fast start.")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 2: print()")

    print('    print("Hello world")')
    print()
    print("  print() takes several arguments and joins them with a space")
    print("  by default:")
    print()
    print('    print("a", "b", "c")')

    print("\n  Running:  ", end="")
    print("a", "b", "c")

    print()
    print("  Two keyword arguments change that behaviour:")
    print()
    print('    print("a", "b", sep="-")     # a-b')
    print('    print("no newline", end="")  # no \\n after this call')
    print()
    print("  Running:  ", end="")
    print("a", "b", sep="-")
    print("  Running:  ", end="")
    print("no newline", end="")
    print("!")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 3: comments and the file that runs itself")

    print("  # starts a comment, to the end of the line -- there is no")
    print("  block-comment syntax the way C has /* */.")
    print()
    print('    # this line does nothing')
    print('    print("this one does")')
    print()
    print("  A script needs no main() -- the file's top-level code is what")
    print("  runs, top to bottom, the moment the interpreter reads it.")

    ui.wait_enter()
    ui.clear_screen()
    ui.exercise(1)

    ui.question(
        'print("a", "b");  -- what appears between them by default?',
        "a space",
        "print() joins its arguments with a single space unless sep= says otherwise.",
    )
    ui.question(
        "Does Python need a separate compile step before running a script?\n"
        "  (answer: yes or no)",
        "no",
        "The interpreter reads and runs the file directly.",
    )

    ui.challenge(
        ["Print your name and the current year, each on its own line."],
        [],
        ["Jose", "2026"],
        [
            'print("Jose")',
            'print("2026")',
        ],
    )

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("SUMMARY")

    print("   - python3 file.py runs directly, no build step")
    print("   - print() joins arguments with a space; sep= and end= change that")
    print("   - a mistake later in the file is only caught once execution reaches it")
    print()
    print("  Module 2: variables, and a division rule that is not C's.")
    ui.wait_enter()


def lesson_02_variables():
    ui.title("MODULE 2 - VARIABLES AND TYPES")

    ui.heading("PART 1: no declaration, and the type can change")

    print("  C makes you say int age = 25 -- the type first, fixed forever.")
    print("  Python just assigns:")
    print()
    print('    age = 25')
    print('    height = 1.75')
    print('    name = "Ana"')
    print()
    print("  The variable is not typeless -- type(age) really is int -- but")
    print("  nothing stops you rebinding the same name to a different type")
    print("  later:")
    print()

    x = 5
    print(f"    x = 5           -> type(x) is {type(x).__name__}")
    x = "now a string"
    print(f'    x = "now..."    -> type(x) is {type(x).__name__}')
    print()
    print("  That is legal because x was never a box with a fixed type. It")
    print("  is a name, and = just points it at a new object.")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 2: two kinds of division")

    a, b = 7, 2
    print("    a, b = 7, 2")
    print()
    print(f"    a / b    -> {a / b}     # true division: always a float")
    print(f"    a // b   -> {a // b}       # floor division: rounds down")
    print()
    print("  / never truncates the way C's int/int does -- 7 / 2 is 3.5,")
    print("  full stop. // is the one that drops the fraction, and it")
    print("  rounds DOWN, not toward zero:")
    print()
    print(f"    -7 // 2  -> {-7 // 2}      # C's -7 / 2 is -3; Python's is -4")
    print(f"    -7 % 2   -> {-7 % 2}       # the remainder follows the divisor's sign")
    print()
    print("  If you know C: this is the one arithmetic rule that is")
    print("  genuinely different, not just spelled differently.")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 3: f-strings")

    name = "Ana"
    pi = 3.14159
    print('    name = "Ana"')
    print("    pi = 3.14159")
    print('    print(f"{name} tem pi={pi:.2f}")')
    print()
    print(f"  Running:  {name} tem pi={pi:.2f}")
    print()
    print("  f before the quote turns {expr} inside the string into its")
    print("  value. :.2f after a colon is a format spec -- two decimal")
    print("  places, rounded for display only; pi itself is unchanged.")

    ui.wait_enter()
    ui.clear_screen()
    ui.exercise(2)

    ui.question("x = 7 / 2   What is x?", "3.5", "/ is true division and never truncates.")
    ui.question(
        "x = 7 // 2   What is x?", "3", "// floors the true result: 3.5 rounds down to 3."
    )
    ui.question(
        "-7 // 2   Does this floor toward zero or toward negative infinity?",
        "negative infinity",
        "Python floors down, always -- -7 // 2 is -4, not -3.",
    )

    ui.challenge(
        ["Declare a = 7 and b = 2, and print their true division", "(should show 3.5)."],
        [],
        ["3.5"],
        ["a = 7", "b = 2", "print(a / b)"],
    )

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("SUMMARY")

    print("   - assignment, not declaration; a name can point at any type")
    print("   - / is always true division; // floors, toward negative infinity")
    print("   - f-strings: f\"{expr:spec}\" -- computed once, formatted for display")
    ui.wait_enter()


def lesson_03_input():
    ui.title("MODULE 3 - READING INPUT")

    ui.heading("PART 1: input() always returns a string")

    print('    name = input("What is your name? ")')
    print()
    print("  Whatever the user types comes back as str, whether it looks")
    print('  like a number or not. Unlike C\'s scanf, there is no size to')
    print("  get wrong -- input() reads the whole line, however long.")
    print()
    print("  For a number, convert explicitly:")
    print()
    print('    age = int(input("Age: "))')
    print('    price = float(input("Price: "))')

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 2: forgetting to convert")

    print("  This is Python's version of a very common beginner bug:")
    print()
    print('    age = input("Age: ")')
    print('    print("Next year: " + age + 1)')
    print()
    print("  Running it, with age typed as 25:")
    print()
    print('    TypeError: can only concatenate str (not "int") to str')
    print()
    print("  + on strings concatenates; + between a str and an int is not")
    print("  defined at all, so Python refuses rather than guess. Verified")
    print("  on 3.14 -- the exact wording has changed across versions, but")
    print("  the refusal has not.")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 3: bad input raises, and can be caught")

    print("    try:")
    print('        n = int(input("Number: "))')
    print('        print(f"You typed {n}")')
    print("    except ValueError:")
    print('        print("That is not a number.")')
    print()
    print("  Typing \"abc\" where a number was expected raises ValueError.")
    print("  Uncaught, that crashes the script with a traceback; caught,")
    print("  you get to say something useful instead. Module 10 covers")
    print("  try/except properly -- this is the shape you will use most.")

    ui.wait_enter()
    ui.clear_screen()
    ui.exercise(3)

    ui.question(
        'age = input("Age: ")   What is the type of age, whatever is typed?',
        "str",
        "input() always returns a string. int()/float() convert it explicitly.",
    )
    ui.question(
        '"Total: " + age + 1  where age is a str  -- what happens?',
        "TypeError",
        "+ between str and int is not defined; Python raises rather than guesses.",
    )

    ui.challenge(
        ["Ask for the user's name and greet them."],
        ["Jose"],
        ["What is your name? Jose", "Hello, Jose"],
        [
            'name = input("What is your name? ")',
            'print(f"Hello, {name}")',
        ],
    )

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("SUMMARY")

    print("   - input() always returns a str; convert with int()/float()")
    print("   - str + int raises TypeError rather than silently doing the wrong thing")
    print("   - int(bad_text) raises ValueError -- catchable with try/except")
    ui.wait_enter()


def lesson_04_conditions():
    ui.title("MODULE 4 - CONDITIONS")

    ui.heading("PART 1: indentation is the block")

    print("  C uses { }. Python has no braces at all -- the indentation")
    print("  itself marks what is inside the if:")
    print()
    print("    score = 85")
    print("    if score >= 90:")
    print('        print("A")')
    print("    elif score >= 80:")
    print('        print("B")')
    print("    else:")
    print('        print("F")')
    print()

    score = 85
    print(f"  Running with score = {score}:  ", end="")
    if score >= 90:
        print("A")
    elif score >= 80:
        print("B")
    else:
        print("F")

    print()
    print("  Get the indentation wrong -- mix tabs and spaces, or indent by")
    print("  a different amount within the same block -- and Python refuses")
    print("  to run at all:")
    print()
    print("    IndentationError: expected an indented block after 'if'")
    print("                       statement on line 1")
    print()
    print("  Verified on 3.14. In C, bad indentation is a style problem the")
    print("  compiler never sees; in Python it is a syntax error, every time.")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 2: truthiness")

    print("  if converts whatever you give it to a bool. These are falsy:")
    print()
    print('    False   0   0.0   ""   []   {}   None')
    print()
    print("  Everything else is truthy, including a string that just says")
    print('  "False" or a list holding a single 0:')
    print()

    print(
        "   ",
        bool(0),
        bool(""),
        bool([]),
        bool(None),
        bool({}),
        "  <- all falsy",
    )
    print("   ", bool(1), bool("x"), bool([0]), " <- all truthy")

    print()
    name = ""
    print('    name = ""')
    print("    if name:")
    print('        print("has name")')
    print("    else:")
    print('        print("no name")')
    print()
    print("  Running:  ", end="")
    if name:
        print("has name")
    else:
        print("no name")

    print()
    print("  An empty string is falsy, so testing `if name:` directly is")
    print('  idiomatic -- there is no need for `if name != "":`.')

    ui.wait_enter()
    ui.clear_screen()
    ui.exercise(4)

    ui.question(
        "Which of these are falsy: 0, 1, [], [0], \"\", \"False\"?\n"
        "  (answer: list just the falsy ones, comma separated)",
        "0, [], \"\"",
        '1, [0] and the string "False" are all truthy -- a non-empty '
        "container or string is truthy regardless of what it contains.",
    )
    ui.question(
        "A block is indented with 2 spaces on one line and 4 on the\n"
        "  next, inside the same if. What happens?",
        "IndentationError",
        "Python does not silently accept inconsistent indentation within one block.",
    )

    ui.challenge(
        ["Read a number and say whether it is positive, negative or zero."],
        ["-5"],
        ["Number: -5", "negative"],
        [
            'n = int(input("Number: "))',
            "if n > 0:",
            '    print("positive")',
            "elif n < 0:",
            '    print("negative")',
            "else:",
            '    print("zero")',
        ],
    )

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("SUMMARY")

    print("   - indentation IS the block; no braces, and inconsistency is a syntax error")
    print("   - 0, 0.0, \"\", [], {}, None are falsy; everything else is truthy")
    print("   - `if name:` is the idiomatic emptiness check")
    ui.wait_enter()


def lesson_05_loops():
    ui.title("MODULE 5 - LOOPS")

    ui.heading("PART 1: for walks a sequence, range() makes one")

    print("  Python's for does not count -- it walks whatever you give it.")
    print("  range() is how you get something to count with:")
    print()
    print("    for i in range(1, 6):")
    print('        print(i, end=" ")')
    print()
    print("  Running:  ", end="")
    for i in range(1, 6):
        print(i, end=" ")
    print()
    print()
    print("  range(1, 6) gives 1, 2, 3, 4, 5 -- the end is EXCLUSIVE, the")
    print("  same rule as C's `i < 5`, just spelled as a boundary instead")
    print("  of a comparison. range(5) alone starts at 0:")
    print()
    print("    for i in range(5):")
    print('        print(i, end=" ")')
    print()
    print("  Running:  ", end="")
    for i in range(5):
        print(i, end=" ")
    print()

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 2: walking a list directly")

    fruits = ["maca", "banana", "cereja"]
    print('    fruits = ["maca", "banana", "cereja"]')
    print("    for f in fruits:")
    print('        print(f"- {f}")')
    print()
    for f in fruits:
        print(f"    - {f}")

    print()
    print("  No index at all, unless you need one -- and when you do,")
    print("  enumerate() gives both without a separate counter variable:")
    print()
    print("    for i, f in enumerate(fruits):")
    print('        print(i, f)')
    print()
    for i, f in enumerate(fruits):
        print(f"    {i} {f}")

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("PART 3: while, break, continue")

    print("    n = 3")
    print("    while n > 0:")
    print('        print(n, end=" ")')
    print("        n -= 1")
    print()
    n = 3
    print("  Running:  ", end="")
    while n > 0:
        print(n, end=" ")
        n -= 1
    print()
    print()
    print("  There is no n-- in Python -- n -= 1 is the whole vocabulary")
    print("  for it. break and continue work exactly as they do in C.")

    ui.wait_enter()
    ui.clear_screen()
    ui.exercise(5)

    ui.question("range(1, 6)  -- is 6 included?", "no", "The end of range() is always exclusive.")
    ui.question(
        "for f in fruits: ...  -- do you need an index to reach each element?\n"
        "  (answer: yes or no)",
        "no",
        "for walks the values directly; enumerate() adds the index only when you ask for it.",
    )

    ui.challenge(
        [
            "Print 1 through 5, one per line. Then print each of",
            '"apple", "banana", "cherry" prefixed with "- ".',
        ],
        [],
        ["1", "2", "3", "4", "5", "- apple", "- banana", "- cherry"],
        [
            "for i in range(1, 6):",
            "    print(i)",
            "",
            'fruits = ["apple", "banana", "cherry"]',
            "for f in fruits:",
            '    print(f"- {f}")',
        ],
    )

    ui.wait_enter()
    ui.clear_screen()
    ui.heading("SUMMARY")

    print("   - for walks a sequence; range(a, b) is a to b-1")
    print("   - looping over a list needs no index; enumerate() adds one when wanted")
    print("   - n -= 1, not n--; break/continue are unchanged from C")
    print()
    print("  Module 6: lists, and the aliasing bug that costs people an afternoon.")
    ui.wait_enter()

File: javascript/src/lessons_basics.js
// Modules 1 to 5 - the basics.
//
// Rule used throughout: nothing is claimed without being shown. Where an
// example can run here, it runs.

const ui = require("./ui.js");

async function lesson01Running() {
  ui.title("MODULE 1 - RUNNING NODE AND CONSOLE.LOG()");

  ui.heading("PART 1: no compile step");

  console.log("  Like Python, JavaScript run through Node has no separate");
  console.log("  build step. Node reads your file and runs it directly:");
  console.log();
  console.log("    node program.js");
  console.log();
  console.log("  That is the whole command. A mistake later in the file is");
  console.log("  only caught once execution actually reaches it -- there is");
  console.log("  no compiler checking the whole file up front the way gcc");
  console.log("  or g++ does.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 2: console.log() and template literals");

  console.log('    console.log("Hello world")');
  console.log('    console.log("a", "b", "c")');
  console.log();
  process.stdout.write("  Running:  ");
  console.log("a", "b", "c");
  console.log();
  console.log("  Several arguments are joined with a space, same as Python's");
  console.log("  print(). For building a string with values inside, use a");
  console.log("  template literal -- backticks, not quotes:");
  console.log();
  console.log('    const name = "Ana";');
  console.log('    console.log(`Hello, ${name}`);');
  console.log();
  const name = "Ana";
  console.log(`  Running: Hello, ${name}`);
  console.log();
  console.log("  ${expr} inside backticks is evaluated and inserted. Regular");
  console.log("  quotes do not do this -- \"Hello, ${name}\" prints literally.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 3: let, const, and no semicolon-or-not debate here");

  console.log("  const for a name you will not reassign, let for one you");
  console.log("  will. Both exist because of var, an older keyword with");
  console.log("  scoping rules that surprise people -- module 7 shows the");
  console.log("  specific way it does. Avoid var in code you write.");
  console.log();
  console.log("  // starts a line comment, same as most languages here.");
  console.log("  There is no main() -- top-level code runs as the file is");
  console.log("  read, same as Python.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.exercise(1);

  await ui.question(
    'console.log("a", "b");  -- what appears between them by default?',
    "a space",
    "console.log() joins its arguments with a single space, same as Python's print()."
  );
  await ui.question(
    'const name = "Ana"; console.log("Hi, ${name}");  -- does ${name} get replaced?\n' +
      "  (answer: yes or no)",
    "no",
    "Regular quotes do not interpolate. Only backtick template literals do."
  );

  await ui.challenge(
    ["Print your name and the current year, each on its own line."],
    [],
    ["Jose", "2026"],
    ['console.log("Jose");', 'console.log("2026");']
  );

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("SUMMARY");

  console.log("   - node file.js runs directly, no build step");
  console.log("   - console.log() joins arguments with a space");
  console.log("   - `${expr}` inside backticks interpolates; regular quotes do not");
  console.log();
  console.log("  Module 2: variables, and a coercion rule Python does not have.");
  await ui.waitEnter();
}

async function lesson02Variables() {
  ui.title("MODULE 2 - VARIABLES AND TYPES");

  ui.heading("PART 1: typeof, and == vs ===");

  console.log("    let age = 25;");
  console.log("    const height = 1.75;");
  console.log();
  let age = 25;
  const height = 1.75;
  console.log(`    typeof age    -> ${typeof age}`);
  console.log(`    typeof height -> ${typeof height}`);
  console.log(`    typeof "x"    -> ${typeof "x"}`);
  console.log();
  console.log("  Two equals operators, and they are not interchangeable:");
  console.log();
  console.log(`    1 == "1"   -> ${1 == "1"}    // converts, then compares`);
  console.log(`    1 === "1"  -> ${1 === "1"}   // compares type and value, no conversion`);
  console.log();
  console.log("  == quietly converts one side to match the other before");
  console.log("  comparing, with rules nobody has memorised:");
  console.log();
  console.log(`    0 == ""    -> ${0 == ""}`);
  console.log(`    0 == "0"   -> ${0 == "0"}`);
  console.log(`    "" == "0"  -> ${"" == "0"}`);
  console.log();
  console.log("  Use === always. It is not a style preference -- == has");
  console.log("  genuine surprises baked into the language specification.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 2: NaN, and a variable's type can change");

  console.log(`    NaN === NaN         -> ${NaN === NaN}`);
  console.log(`    Number.isNaN(NaN)   -> ${Number.isNaN(NaN)}`);
  console.log();
  console.log("  NaN famously does not equal itself. Number.isNaN() is the");
  console.log("  correct check; === against NaN directly is always false.");
  console.log();
  age = "now a string";
  console.log("    age = 25;");
  console.log('    age = "now a string";    // legal -- let rebinds freely');
  console.log(`    typeof age -> ${typeof age}`);
  console.log();
  console.log("  const stops REASSIGNMENT, not mutation, and has nothing to");
  console.log("  say about type -- module 6 covers what const does and does");
  console.log("  not protect on an array or object.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 3: floating point, unchanged from every other language here");

  console.log(`    0.1 + 0.2  -> ${0.1 + 0.2}`);
  console.log();
  console.log("  Same binary floating point as C, Python and everything");
  console.log("  else here. Not a JavaScript bug -- for money, work in whole");
  console.log("  cents, same rule as always.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.exercise(2);

  await ui.question('1 == "1"   Does this compare equal?', "yes", "== converts types before comparing.");
  await ui.question(
    '1 === "1"   Does this compare equal?',
    "no",
    "=== never converts. Different types are never equal under ===."
  );
  await ui.question("NaN === NaN   Equal?", "no", "NaN never equals anything, including itself. Use Number.isNaN().");

  await ui.challenge(
    ["Declare text = \"5\" and number = 5. Print typeof text,", "typeof number, then text == number, then text === number."],
    [],
    ["string number", "true false"],
    [
      'const text = "5";',
      "const number = 5;",
      "console.log(typeof text, typeof number);",
      "console.log(text == number, text === number);",
    ]
  );

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("SUMMARY");

  console.log("   - always ===, never == -- == converts with rules nobody remembers");
  console.log("   - NaN !== NaN; check with Number.isNaN()");
  console.log("   - let/const control reassignment, not type or mutability");
  await ui.waitEnter();
}

async function lesson03Input() {
  ui.title("MODULE 3 - READING INPUT");

  ui.heading("PART 1: reading a line is asynchronous");

  console.log("  C's scanf, Python's input(), even C++'s cin -- all block");
  console.log("  until a line arrives. Node has nothing built in that does");
  console.log("  that. Reading input is an operation that takes time, and");
  console.log("  Node is built around never blocking while waiting -- so");
  console.log("  reading a line is asynchronous, and you await it:");
  console.log();
  console.log('    const readline = require("node:readline");');
  console.log('    const rl = readline.createInterface({ input: process.stdin });');
  console.log("    const it = rl[Symbol.asyncIterator]();");
  console.log();
  console.log("    async function ask(prompt) {");
  console.log("        process.stdout.write(prompt);");
  console.log("        const { value } = await it.next();");
  console.log("        return value;");
  console.log("    }");
  console.log();
  console.log("  Every function that calls ask() must itself be async, and");
  console.log("  every caller of THAT function must await it -- async is");
  console.log("  contagious upward through the whole call chain, all the");
  console.log("  way to the top-level code that starts everything.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 2: try it, for real");

  const name = await ui.ask("\n  What is your name? ");
  console.log(`\n  Hello, ${name}. This course's own ui.js is the ask()`);
  console.log("  function shown above, reading your actual keyboard.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 3: a real gotcha, specific to testing non-interactively");

  console.log("  If you test a script by piping input instead of typing it:");
  console.log();
  console.log('    printf "a\\nb\\n" | node script.js');
  console.log();
  console.log("  and the script asks two questions with the readline/promises");
  console.log("  library's rl.question(), the SECOND question can silently");
  console.log("  hang forever with no error. Verified: it happens because");
  console.log("  printf closes the pipe the instant it finishes writing, and");
  console.log("  once stdin has already hit that end-of-input, a second");
  console.log("  pending question never resolves.");
  console.log();
  console.log("  A real terminal never closes stdin while you are sitting");
  console.log("  there, so this never affects the course itself. It does");
  console.log("  affect testing a script non-interactively -- if you pipe");
  console.log("  input to test something you wrote, and it hangs on the");
  console.log("  second prompt, this is why.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.exercise(3);

  await ui.question(
    "Does Node have a built-in function that blocks and returns a\n" +
      "  typed line the way Python's input() does?  (answer: yes or no)",
    "no",
    "Reading a line is asynchronous in Node; there is no blocking equivalent built in."
  );
  await ui.question(
    "If function b() awaits an async function a(), can b() itself\n" +
      "  be a plain, non-async function?  (answer: yes or no)",
    "no",
    "await only works inside an async function -- the requirement propagates upward " +
      "through every caller."
  );

  await ui.challenge(
    ["Ask for the user's name and greet them."],
    ["Jose"],
    ["What is your name? Jose", "Hello, Jose"],
    [
      'const readline = require("node:readline");',
      "const rl = readline.createInterface({ input: process.stdin });",
      "const it = rl[Symbol.asyncIterator]();",
      "",
      "async function main() {",
      '    process.stdout.write("What is your name? ");',
      "    const { value: name } = await it.next();",
      "    console.log(`Hello, ${name}`);",
      "    rl.close();",
      "}",
      "",
      "main();",
    ]
  );

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("SUMMARY");

  console.log("   - reading input is async; every caller up the chain must be too");
  console.log("   - a real terminal is fine; a pre-closed test pipe can hang a second read");
  console.log("   - this course's own ask() avoids that by draining readline's iterator by hand");
  await ui.waitEnter();
}

async function lesson04Conditions() {
  ui.title("MODULE 4 - CONDITIONS");

  ui.heading("PART 1: if / else if / else");

  console.log("    const score = 85;");
  console.log("    if (score >= 90) {");
  console.log('        console.log("A");');
  console.log("    } else if (score >= 80) {");
  console.log('        console.log("B");');
  console.log("    } else {");
  console.log('        console.log("F");');
  console.log("    }");
  console.log();
  const score = 85;
  process.stdout.write(`  Running with score = ${score}:  `);
  if (score >= 90) {
    console.log("A");
  } else if (score >= 80) {
    console.log("B");
  } else {
    console.log("F");
  }
  console.log();
  console.log("  Braces mark the block, same as C. Only the first true");
  console.log("  branch runs, so order matters, same rule as everywhere else.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 2: truthy and falsy -- and this is where JS and Python disagree");

  console.log("  Falsy in JavaScript: false, 0, \"\", null, undefined, NaN.");
  console.log("  So far, similar to Python. Here is where they diverge:");
  console.log();
  console.log(`    Boolean([])   -> ${Boolean([])}`);
  console.log(`    Boolean({})   -> ${Boolean({})}`);
  console.log();
  console.log("  TRUE. An empty array and an empty object are both truthy in");
  console.log("  JavaScript -- the opposite of Python, where [] and {} are");
  console.log("  falsy. If you know Python: `if (arr)` is never a length");
  console.log("  check here. Use `if (arr.length)` or `if (arr.length === 0)`.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 3: = is still not ==, and still just a warning at best");

  console.log("    let x = 3;");
  console.log("    if (x = 5) { ... }");
  console.log();
  console.log("  Assigns 5, evaluates to 5, 5 is truthy, the block runs.");
  console.log("  Verified: this executes with no error at all -- unlike gcc");
  console.log("  and g++, Node does not even warn about it by default.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.exercise(4);

  await ui.question("Boolean([])  -- truthy or falsy?", "truthy", "Empty arrays and objects are truthy in JS, unlike Python.");
  await ui.question(
    "To check if an array is empty, do you write `if (arr)` or\n  `if (arr.length === 0)`?",
    "if (arr.length === 0)",
    "if (arr) is always true for an array, empty or not -- arrays are objects, and objects are truthy."
  );

  await ui.challenge(
    ["Read a number and say whether it is positive, negative or zero."],
    ["-5"],
    ["Number: -5", "negative"],
    [
      "const readline = require(\"node:readline\");",
      "const rl = readline.createInterface({ input: process.stdin });",
      "const it = rl[Symbol.asyncIterator]();",
      "",
      "async function main() {",
      '    process.stdout.write("Number: ");',
      "    const { value } = await it.next();",
      "    const n = Number(value);",
      "",
      "    if (n > 0) console.log(\"positive\");",
      "    else if (n < 0) console.log(\"negative\");",
      "    else console.log(\"zero\");",
      "    rl.close();",
      "}",
      "",
      "main();",
    ]
  );

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("SUMMARY");

  console.log("   - only the first true branch runs; order matters");
  console.log("   - [] and {} are TRUTHY in JS -- opposite of Python");
  console.log("   - = still is not ==, and Node does not even warn about it");
  await ui.waitEnter();
}

async function lesson05Loops() {
  ui.title("MODULE 5 - LOOPS");

  ui.heading("PART 1: for, while, unchanged from C");

  console.log("    for (let i = 1; i <= 5; i++) {");
  console.log('        process.stdout.write(i + " ");');
  console.log("    }");
  console.log();
  process.stdout.write("  Running:  ");
  for (let i = 1; i <= 5; i++) {
    process.stdout.write(i + " ");
  }
  console.log();
  console.log();
  console.log("  Same three-part shape, same off-by-one risk with <= where");
  console.log("  < was meant. `let`, not `var`, for the counter -- module 7");
  console.log("  covers exactly why that choice matters inside a loop.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 2: for...of and for...in are not the same loop");

  const fruits = ["apple", "banana", "cherry"];
  console.log('    const fruits = ["apple", "banana", "cherry"];');
  console.log("    for (const f of fruits) console.log(f);      // VALUES");
  console.log("    for (const i in fruits) console.log(i);      // KEYS, as strings");
  console.log();
  for (const f of fruits) {
    console.log(`    - ${f}`);
  }
  console.log();
  for (const i in fruits) {
    console.log(`    in: ${i}  (${typeof i})`);
  }
  console.log();
  console.log("  for...in gives the indices, as strings -- \"0\", \"1\", \"2\", not");
  console.log("  the elements. Using for...in on an array is almost always a");
  console.log("  mistake; remember it as: of for values, in for keys.");

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("PART 3: while");

  console.log("    let n = 3;");
  console.log("    while (n > 0) {");
  console.log('        process.stdout.write(n + " ");');
  console.log("        n--;");
  console.log("    }");
  console.log();
  let n = 3;
  process.stdout.write("  Running:  ");
  while (n > 0) {
    process.stdout.write(n + " ");
    n--;
  }
  console.log();

  await ui.waitEnter();
  ui.clearScreen();
  ui.exercise(5);

  await ui.question("for (const x of arr)  -- does x hold values or indices?", "values", "of walks the values directly.");
  await ui.question("for (const x in arr)  -- does x hold values or indices?", "indices", "in walks the keys/indices, as strings.");

  await ui.challenge(
    [
      "Print 1 through 5, one per line. Then print each of",
      '"apple", "banana", "cherry" prefixed with "- ".',
    ],
    [],
    ["1", "2", "3", "4", "5", "- apple", "- banana", "- cherry"],
    [
      "for (let i = 1; i <= 5; i++) {",
      "    console.log(i);",
      "}",
      "",
      'const fruits = ["apple", "banana", "cherry"];',
      "for (const f of fruits) {",
      '    console.log("- " + f);',
      "}",
    ]
  );

  await ui.waitEnter();
  ui.clearScreen();
  ui.heading("SUMMARY");

  console.log("   - for/while carry over unchanged, off-by-one included");
  console.log("   - of -> values, in -> keys as strings; do not mix them up on arrays");
  console.log();
  console.log("  Module 6: arrays, and the same aliasing bug Python has too.");
  await ui.waitEnter();
}

module.exports = {
  lesson01Running,
  lesson02Variables,
  lesson03Input,
  lesson04Conditions,
  lesson05Loops,
};
